src/MassTransit.Abstractions/Internals/Extensions/TaskExtensions.cs

```diff
@@ -105,17 +105,32 @@ async Task<T> WaitAsync()
 
             async Task WaitAsync()
             {
-                var delayTask = Task.Delay(Debugger.IsAttached ? Timeout.InfiniteTimeSpan : timeout, cancellationToken);
-
-                var completed = await Task.WhenAny(task, delayTask).ConfigureAwait(false);
-                if (completed == delayTask)
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                try
                 {
-                    task.IgnoreUnobservedExceptions();
+                    Task delayTask = Task.Delay(Debugger.IsAttached ? Timeout.InfiniteTimeSpan : timeout, cts.Token)
+                        .ContinueWith(t => t, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
 
-                    throw new TimeoutException(FormatTimeoutMessage(memberName, filePath, lineNumber));
-                }
+                    var completed = await Task.WhenAny(task, delayTask).ConfigureAwait(false);
 
-                task.GetAwaiter().GetResult();
+                    cts.Cancel();
+
+                    if (completed == delayTask)
+                    {
+                        task.IgnoreUnobservedExceptions();
+
+                        throw new TimeoutException(FormatTimeoutMessage(memberName, filePath, lineNumber));
+                    }
+
+                    if (!delayTask.IsCompleted)
+                        await delayTask.ConfigureAwait(false);
```

Comment:
No need to await that task. It's canceled so the runtime will do the right job, and the `await` here is just an artificial bloat of code.

Reply:
it is awaited only if it is not completed, if it is completed just pass, and above the point of using ContinueWith to avoid throwing exception here